### PR TITLE
hidraw: Implement hid_error()

### PR DESCRIPTION
**Pull request by z3ntu**
Here is my first try at implementing error messages for the Linux/hidraw implementation.

* I used a thread-local variable for non-device error messages.
* Non-device errors can be obtained by passing `NULL` into `hid_error` as I suggested in #123.
* I **belive** to have covered all system calls that could fail; please check this.

A few extra questions:

* I am not sure why in `utf8_to_wchar_t` the input is called `utf8`. As far as I know, the encoding used in `mbstowcs` depends on the current locale, if if I haven't set my locale to use UTF8, then the input is not UTF8.
* The `register_error` equivalent in `windows/hid.c` doesn't use its argument `const char *op`. Is that intended?
* In the Windows implementation, do functions like [`HidD_GetPreparsedData`](http://msdn.microsoft.com/en-us/library/windows/hardware/ff539679%28v=vs.85%29.aspx) actually set the error code for `GetLastError()`?

Originally https://github.com/signal11/hidapi/pull/125

### CODE REVIEW

**Comment by Youw** [comment]
When a client code gets an "error string" as `char*` (or, in our case `wchar_t*`), it likely assumes, that that is a valid c-string, and for instance try to:
```cpp
printf("Last HIDAPI call status: %S", hide_error());
```
and that would immediately lead to crash, since `hide_error()` sometime returns `NULL`.

I'd suggest always return a valid c-string (non-NULL) instead.
If there is no error, we might return `""` or `"Success"`, or smth like that.

**Comment by Youw** [comment]
my suggestion stands

@z3ntu if you accept - please update PR accordingly, otherwise give your note why you wouldn't do it within this PR; maybe we'll have a separate task for it

**Comment by z3ntu** [comment]
Yes, will update it. I've just been travelling a lot in the past couple of days.

Do you think it's better that if the `register_global_error` function receives a NULL argument, that it sets the `last_global_error_str` variable to `Success` or that the `register_global_error` function should be called with parameter `Success`? I'm personally for the first option.

**Comment by Youw** [comment]
- first looks OK
- `register_global_success` is an alternative
- `hid_error` could return "Success" if `last_global_error_str` is NULL

to be honest I'm more concerned of usage on the user side, and I'll support any reasonable good implementation, so it is up to you to decide
(and yes, I think passing "Success" in each place wouldn't look very nice)

**Comment by z3ntu** [comment]
The last sounds the best imo

File: linux/hid.c [hunk 1]
```diff
@@ -639,15 +690,19 @@ hid_device * HID_API_EXPORT hid_open_path(const char *path)
 
 		/* Get Report Descriptor Size */
 		res = ioctl(dev->device_handle, HIDIOCGRDESCSIZE, &desc_size);
-		if (res < 0)
-			perror("HIDIOCGRDESCSIZE");
-
+		if (res < 0) {
+			char errormsg[100];
+			snprintf(errormsg, sizeof(errormsg), "ioctl (GRDESCSIZE): %s", strerror(errno));
+			register_device_error(dev, errormsg);
```
Comment (z3ntu) [inline linux/hid.c]: I'm not very fond of this construct being present in multiple places. Should I create a helper function for that? If yes, name suggestions?

File: linux/hid.c [hunk 2]
```diff
@@ -639,15 +690,19 @@ hid_device * HID_API_EXPORT hid_open_path(const char *path)
...
+		}
 
 		/* Get Report Descriptor */
 		rpt_desc.size = desc_size;
 		res = ioctl(dev->device_handle, HIDIOCGRDESC, &rpt_desc);
 		if (res < 0) {
-			perror("HIDIOCGRDESC");
+			char errormsg[100];
+			snprintf(errormsg, sizeof(errormsg), "ioctl (GRDESC): %s", strerror(errno));
```
Comment (z3ntu) [inline linux/hid.c]: I've taken the opportunity to also improve the messages here to match the ioctl messages further below :)

**Comment by z3ntu** [comment]
Should be good for a review now

File: linux/hid.c (hunk 1 above)
Comment (Youw) [reply linux/hid.c]: `register_device_ioctl_error`?

File: linux/hid.c [hunk 3]
```diff
@@ -269,7 +303,7 @@ static int get_device_string(hid_device *dev, enum device_string_id key, wchar_t
 	/* Create the udev object */
 	udev = udev_new();
 	if (!udev) {
-		printf("Can't create udev\n");
+		register_global_error("Couldn't create udev context\n");
```
Comment (Youw) [inline linux/hid.c]:
```suggestion
		register_global_error("Couldn't create udev context");
```

File: linux/hid.c [hunk 4]
```diff
@@ -395,7 +431,7 @@ struct hid_device_info  HID_API_EXPORT *hid_enumerate(unsigned short vendor_id,
...
```
Comment (Youw) [inline linux/hid.c]:
```suggestion
		register_global_error("Couldn't create udev context");
```

File: windows/hid.c [hunk 5]
```diff
@@ -881,7 +881,11 @@ int HID_API_EXPORT_CALL HID_API_CALL hid_get_indexed_string(hid_device *dev, int
 
 HID_API_EXPORT const wchar_t * HID_API_CALL  hid_error(hid_device *dev)
 {
-	return (wchar_t*)dev->last_error_str;
+	if (dev)
+		return (wchar_t*)dev->last_error_str;
```
Comment (Youw) [inline windows/hid.c]: `L"Success"` if NULL here too

File: linux/hid.c (hunk 1 above)
Comment (z3ntu) [reply linux/hid.c]: I've gone for `static void register_device_error_format(hid_device *dev, const char *format, ...)`

File: linux/hid.c [hunk 6]
```diff
@@ -130,6 +135,48 @@ static wchar_t *utf8_to_wchar_t(const char *utf8)
...
+	va_list args;
+	va_start(args, format);
+
+	char msg[100];
+	vsnprintf(msg, sizeof(msg), format, args);
+
+	va_end(args);
+
+	register_device_error(dev, msg);
+}
```
Comment (z3ntu) [inline linux/hid.c]: Can someone confirm that this C code is written correctly and I didn't make some fundamental mistake with how the `char msg[100]` is handled? :D

Comment (Youw) [reply linux/hid.c]: LGTM

**Comment by Youw** [comment]
@jdk, @todbot, @Qbicz your comments?

File: linux/hid.c [hunk 7]
```diff
@@ -130,6 +135,48 @@ static wchar_t *utf8_to_wchar_t(const char *utf8)
...
+ * The given error message will be copied (and decoded according to the
+ * currently locale, so do not pass in string constants).
+ * The last stored global error message is freed.
+ * Use register_device_error(device, NULL) to indicate "no error". */
+static void register_device_error(hid_device *dev, const char *msg)
+{
+	if (dev->last_error_str)
+		free(dev->last_error_str);
+
+	dev->last_error_str = msg? utf8_to_wchar_t(msg): NULL;
```
Comment (Qbicz) [inline linux/hid.c]: `msg` will be evaluated to `false` when you call `register_device_error(NULL);` so I think you could change
```
dev->last_error_str = msg? utf8_to_wchar_t(msg): NULL;
```
to 
```
dev->last_error_str = msg;
```

Comment (Qbicz) [reply linux/hid.c]: Sorry, I forgot about need to call utf8_to_wchar_t. So the `? :` is needed.

File: linux/hid.c (hunk 6 above)
Comment (Qbicz) [reply linux/hid.c]: You pass the pointer to local `msg[100]` which is on stack, but in `register_device_error()` it is copied to `dev->last_error_str` which is dynamically allocated and it's lifespan is the same as lifespan of a device. This is copied before the stack is freed. So it's fine ;)

Comment (z3ntu) [reply linux/hid.c]: Great! :)

File: linux/hid.c (hunk 7 above)
Comment (z3ntu) [reply linux/hid.c]: Applied
```diff
diff --git a/linux/hid.c b/linux/hid.c
index 6ce2d9e..4b6cad3 100644
--- a/linux/hid.c
+++ b/linux/hid.c
@@ -146,7 +146,7 @@ static void register_global_error(const char *msg)
        if (last_global_error_str)
                free(last_global_error_str);
 
-       last_global_error_str = msg? utf8_to_wchar_t(msg): NULL;
+       last_global_error_str = utf8_to_wchar_t(msg);
 }
 
 
@@ -160,7 +160,7 @@ static void register_device_error(hid_device *dev, const char *msg)
        if (dev->last_error_str)
                free(dev->last_error_str);
 
-       dev->last_error_str = msg? utf8_to_wchar_t(msg): NULL;
+       dev->last_error_str = utf8_to_wchar_t(msg);
 }
 
 /* See register_device_error, but you can pass a format string into this function. */
```

Comment (Youw) [reply linux/hid.c]: looks good

File: hidapi/hidapi.h [hunk 8]
```diff
@@ -378,8 +394,18 @@ extern "C" {
 
 		/** @brief Get a string describing the last error which occurred.
 
+			Whether a function sets the last error is noted in its
+			documentation. These functions will reset the last error
+			to NULL before their execution.
+
+			Strings returned from hid_error() must not be freed by the user!
+
+			This function is thread-safe, and error messages are thread-local.
```
Comment (Youw) [inline hidapi/hidapi.h]: And it is not thread-safe, when you're using same device from different threads.

**Pull request merged**